src/cpp/session/modules/SessionUserCommands.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+ * SessionUserCommands.cpp
+ *
+ * Copyright (C) 2009-12 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "SessionUserCommands.hpp"
+
+#include <boost/bind.hpp>
+
+#include <core/Error.hpp>
+#include <core/Exec.hpp>
+
+#include <r/RExec.hpp>
+#include <r/RSexp.hpp>
+#include <r/RRoutines.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace user_commands {
+
+using namespace rstudio::core;
+
+namespace {
+
+json::Array jsonFromUserCommandResult(SEXP resultSEXP)
+{
+   json::Array jsonData;
```

Comment:
We have a family of type-checking functions in r::sexp (e.g. isString, isLanguage, etc.). Could we add one for this case and use it here?